File: src/0229-Majority-Element-II/0229.py
class Solution:
    def majorityElement(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        if not nums:
            return []

        c1, c2, m, n = 0, 0, None, None
        for num in nums:
            if num == m:
                c1 += 1
            elif num == n:
                c2 += 1
            elif c1 == 0:
                m, c1 = num, 1
            elif c2 == 0:
                n, c2 = num, 1
            else:
                c1 -= 1
                c2 -= 1

        return [num for num in set([m, n]) \
                    if nums.count(num) > len(nums)//3]
        

if __name__ == "__main__":
    nums = [0, -1, 2, 2, -1]
    print(Solution().majorityElement(nums))

File: src/0229-Majority-Element-II/0229.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> Vec<i32> {
        nums.iter()
            .fold([(0, 0), (0, 0)], |arr, &i| {
                match arr {
                    [(e, c), d] if e == i => [(e, c + 1), d],
                    [d, (e, c)] if e == i => [d, (e, c + 1)],
                    [(_, c), d] if c == 0 => [(i, 1), d],
                    [d, (_, c)] if c == 0 => [d, (i, 1)],
                    [(e1, c1), (e2, c2)] => [(e1, c1 - 1), (e2, c2 - 1)],             
                }
            })
            .iter()
            .filter(|x| x.1 > 0 && nums.iter().filter(|&i| *i == x.0).count() > nums.len() / 3)
            .map(|x| x.0)
            .collect()
    }
}